File: OrcScala/src/orc/lib/util/Random.java
//
// Random.java -- Java class Random
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2013 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.util;

import java.math.BigInteger;

import orc.error.runtime.ArityMismatchException;
import orc.error.runtime.TokenException;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.PartialSite;
import orc.values.sites.compatibility.Types;

public class Random extends PartialSite implements TypedSite {

	java.util.Random rnd;

	public Random() {
		rnd = new java.util.Random();
	}

	@Override
	public Object evaluate(final Args args) throws TokenException {
		if (args.size() > 1) {
			throw new ArityMismatchException(1, args.size());
		}
		if (args.size() == 0) {
			return BigInteger.valueOf(rnd.nextInt());
		}

		if (!(args.getArg(0) instanceof Number) ||
		    args.longArg(0) > Integer.MAX_VALUE ||
		    args.longArg(0) <= 0L || ((Number) args.getArg(0)).doubleValue() > Integer.MAX_VALUE ||
		    ((Number) args.getArg(0)).doubleValue() <= 0.0 ||
		    ((Number) args.getArg(0)).doubleValue() != Math.rint(((Number) args.getArg(0)).doubleValue()) ) {
			throw new IllegalArgumentException("Random's argument must be an integer strictly between 0 and 2**31");
		}

		return BigInteger.valueOf(rnd.nextInt(args.intArg(0)));
	}

	@Override
	public Type orcType() {
		return Types.overload(Types.function(Types.integer()), Types.function(Types.integer(), Types.integer()));
	}

}

File: OrcScala/src/orc/lib/util/Random.scala
//
// Random.scala -- Orc methods Random and URandom
// Project OrcScala
//
// Copyright (c) 2018 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.util

import java.math.BigInteger
import java.util.concurrent.ThreadLocalRandom

import orc.{ DirectInvoker, OrcRuntime }
import orc.values.sites.{ DirectSite, FunctionalSite, IllegalArgumentInvoker, LocalSingletonSite, Range, SiteMetadata }

object Random extends DirectSite with SiteMetadata with FunctionalSite with Serializable with LocalSingletonSite {
  class ArgInvoker extends DirectInvoker {
    def canInvoke(target: AnyRef, arguments: Array[AnyRef]): Boolean = {
      target == Random && arguments.length == 1 && arguments(0).isInstanceOf[Number]
    }

    def invokeDirect(target: AnyRef, arguments: Array[AnyRef]): AnyRef = {
      orc.run.StopWatches.implementation {
        val n = arguments(0).asInstanceOf[Number]
        if (n.longValue() <= Int.MaxValue)
          ThreadLocalRandom.current().nextInt(n.intValue()).asInstanceOf[AnyRef]
        else
          throw new IllegalArgumentException(s"$Random($n): bound much be less than 2**31.")
      }
    }
  }
  class NoArgInvoker extends DirectInvoker {
    def canInvoke(target: AnyRef, arguments: Array[AnyRef]): Boolean = {
      target == Random && arguments.length == 0
    }

    def invokeDirect(target: AnyRef, arguments: Array[AnyRef]): AnyRef = {
      orc.run.StopWatches.implementation {
        ThreadLocalRandom.current().nextInt().asInstanceOf[AnyRef]
      }
    }
  }

  def getInvoker(runtime: OrcRuntime, args: Array[AnyRef]) = {
    args.length match {
      case 0 => new NoArgInvoker()
      case 1 =>
        args(0) match {
          case _: BigInt | _: BigInteger | _: Integer | _: java.lang.Long =>
            new ArgInvoker()
          case _ =>
            IllegalArgumentInvoker(this, args)
        }
      case _ =>
        IllegalArgumentInvoker(this, args)
    }
  }

  override def publications: Range = Range(0, 1)
}

object URandom extends DirectSite with SiteMetadata with FunctionalSite with Serializable with LocalSingletonSite {
  class NoArgInvoker extends DirectInvoker {
    def canInvoke(target: AnyRef, arguments: Array[AnyRef]): Boolean = {
      target == URandom && arguments.length == 0
    }

    def invokeDirect(target: AnyRef, arguments: Array[AnyRef]): AnyRef = {
      orc.run.StopWatches.implementation {
        ThreadLocalRandom.current().nextDouble().asInstanceOf[AnyRef]
      }
    }
  }

  def getInvoker(runtime: OrcRuntime, args: Array[AnyRef]) = {
    args.length match {
      case 0 => new NoArgInvoker()
      case _ =>
        IllegalArgumentInvoker(this, args)
    }
  }

  override def publications: Range = Range(0, 1)
}

